zk_toolbox/crates/config/src/explorer_compose.rs

```diff
@@ -0,0 +1,243 @@
+            config.services.worker_http_port,
+            config.services.data_fetcher_http_port,
+            l2_rpc_url.as_ref(),
+            &db_url,
+            config.services.get_batches_processing_polling_interval(),
+        )
+        .context("Failed to create worker service")?;
+        services.insert(Self::worker_name(chain_name), worker);
+
+        Ok(Self {
+            docker_compose: DockerComposeConfig { services },
+        })
+    }
+
+    fn create_api_service(chain_name: &str, port: u16, db_url: &str) -> DockerComposeService {
+        DockerComposeService {
+            image: EXPLORER_API_DOCKER_IMAGE.to_string(),
+            platform: Some("linux/amd64".to_string()),
+            ports: Some(vec![format!("{}:{}", port, port)]),
+            volumes: None,
+            depends_on: Some(vec![Self::worker_name(chain_name)]),
+            restart: None,
+            environment: Some(HashMap::from([
+                ("PORT".to_string(), port.to_string()),
+                ("LOG_LEVEL".to_string(), "verbose".to_string()),
+                ("NODE_ENV".to_string(), "development".to_string()),
+                ("DATABASE_URL".to_string(), db_url.to_string()),
+            ])),
+            extra_hosts: Some(vec!["host.docker.internal:host-gateway".to_string()]),
+        }
+    }
+
+    fn create_data_fetcher_service(port: u16, l2_rpc_url: &str) -> DockerComposeService {
+        DockerComposeService {
+            image: EXPLORER_DATA_FETCHER_DOCKER_IMAGE.to_string(),
+            platform: Some("linux/amd64".to_string()),
+            ports: Some(vec![format!("{}:{}", port, port)]),
+            volumes: None,
+            depends_on: None,
+            restart: None,
+            environment: Some(HashMap::from([
+                ("PORT".to_string(), port.to_string()),
+                ("LOG_LEVEL".to_string(), "verbose".to_string()),
+                ("NODE_ENV".to_string(), "development".to_string()),
+                ("BLOCKCHAIN_RPC_URL".to_string(), l2_rpc_url.to_string()),
+            ])),
+            extra_hosts: Some(vec!["host.docker.internal:host-gateway".to_string()]),
+        }
+    }
+
+    fn create_worker_service(
+        chain_name: &str,
+        port: u16,
+        data_fetcher_port: u16,
+        l2_rpc_url: &str,
+        db_url: &Url,
+        batches_processing_polling_interval: u64,
+    ) -> anyhow::Result<DockerComposeService> {
+        let data_fetcher_url = format!(
+            "http://{}:{}",
+            Self::data_fetcher_name(chain_name),
+            data_fetcher_port
+        );
+
+        // Parse database URL
+        let db_config = db::DatabaseConfig::from_url(db_url)?;
+        let db_user = db_url.username().to_string();
+        let db_password = db_url.password().unwrap_or("");
+        let db_port = db_url.port().unwrap_or(5432);
+        let db_host = db_url
+            .host_str()
+            .context("Failed to parse database host")?
+            .to_string();
```

Comment:
I kinda wanted to move that to DatabaseConfig as methods but was not sure that logic for parsing was good enough: e.g. "postgres://postgres@localhost:5432/database_name?password=notsecurepassword" is also valid url, but by this method it won't be parsed.
I wanted to parse with PgConnectOptions, but the one I need (with FromStr) is from [sqlx_postgres](https://docs.rs/sqlx-postgres/latest/sqlx_postgres/struct.PgConnectOptions.html) and not from sqlx. Anyway, let me know if we have a good way to parse postgres url.

Reply:
Unfortunately, that's how it's configured in block-explorer worker: [[1](https://github.com/matter-labs/block-explorer/blob/main/docker-compose-cli.yaml#L18-L21)], [[2](https://github.com/matter-labs/block-explorer/blob/d951ad07711b36cf52eb4d5099e33ff8bcc04705/packages/worker/README.md?plain=1#L22-L25)], [[3](https://github.com/matter-labs/block-explorer/blob/d951ad07711b36cf52eb4d5099e33ff8bcc04705/packages/worker/src/typeorm.config.ts#L8-L12)].